### PR TITLE
Allow squid tuning in proxy containers

### PR DESCRIPTION
**Pull request by avshiliaev**
## What does this PR change?

Allows squid tuning in proxy containers.

- [ ] Limit bandwidth for squid

## GUI diff

Before:

After:

Screenshot_20230428_101351
Screenshot_20230428_101416

- [x] **DONE**

## Documentation

- [ ] **DONE**

## Test coverage
- Unit tests were added

- [ ] **DONE**

## Links

Implements https://github.com/SUSE/spacewalk/issues/20642

- [ ] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by avshiliaev** [comment]
Hey @aaannz @cbosdo, there's my first take on how to allow proxy squid configuration. 

### Overview

* There's the field called `bandwidth_limit_kbps` in the `config.yaml`. It's added when the config is generated through the UI (see the screenshots above).
* The field is a positive integer, so if it's greater then 0, then the limit is set via the [delay pool](https://wiki.squid-cache.org/Features/DelayPools). 
* The delay pool config is a separate file and the entire `conf.d` dir [is added to the config](https://github.com/uyuni-project/uyuni/pull/6955/files#diff-786076059f6e2314df4d85db59705a8557a3d6531d6c2e51177c4277649dee78). 
* Whenever the container starts, the `containers/proxy-squid-image/uyuni-configure.py` script gets executed. It's setting the `/etc/squid/conf.d/delay.conf` depending on the proxy configuration `config.yaml`. [See here](https://github.com/uyuni-project/uyuni/pull/6955/files#diff-cfdad0aebaf1915e193f498d8ad1da4e1c0ccc4186a1c161448d3046072d6f10). **So basically we need a clean restart, so the container gets newly recreated and the script sets the config properly.**
* for the systemd case there's an extension of the `containers/proxy-systemd-services/uyuni-proxy.sh ` [which is first mentioned here](https://documentation.suse.com/suma/4.3/en/suse-manager/installation-and-upgrade/proxy-container-installation.html#_using_a_custom_container_image_for_a_service). [It can replace](https://github.com/uyuni-project/uyuni/pull/6955/files#diff-ac8ef6e888c2587bbf36be5bcd625bb5e2b44dcdf9dca14cf588eda906c87b33) the `bandwidth_limit_kbps` value in the `/etc/uyuni/proxy/config.yaml` and restart the service. 
* for the helm case there's just a field in the [ConfigMap](https://github.com/uyuni-project/uyuni/pull/6955/files#diff-5379c71dbba3acab8b256361e59e3c842898c59be1a08b25d5ad9a15177ab381) which reads it as usual from the `config.yaml` generated by the server. It can be changed manually. 

### Now the questions

What do we want to do with the salt here? We can create an sls to call either the `uyuni-proxy.sh` or the `kubectl` based on the container runtime. 

Something along these lines 

```yaml
update_squid_config:
{%- if pillar['custom_info']["runtime"] == "podman" %}
  cmd.run:
    - name: uyuni-proxy.sh update config -l 256
{%- else %}
  cmd.run:
    - name: ...
{%- endif %}
```

How do we know the actual runtime? Should the user set in up in the custom_info pillars, or should we have a kind of grains to detect the runtime on the host? Where should we place the code and stuff?

**Comment by aaannz** [comment]
This is completely different that what I had in mind. 😄 

1) This rate limit is specific to one user and I am not sure how useful it is generally. So I did not think to add this to the GUI itself. However modifying squid config is something more users would want.
So what was I thinking - if you want to have it in GUI or proxy config generally, then customer will create their own squid config snippet, which we store to the proxy config tarball and extract to some well known space as and provide it as a squid volume mounted to `/etc/squid/conf.d`.
Goal IMO is to have sane squid defaults which customers can expand using their snippets. If there is a problem, we can ask them to use default config to reproduce a problem.
In the issue I mentioned we need to be careful with ACLs and similar because how squid evaluates them and it is easy for customer to get some unexpected behavior (see [squid acl docs](http://wiki.squid-cache.org/SquidFaq/SquidAcl#andor-logic).

2) I think we can disregard this completely.
In the issue they adapt their zypper config on the proxy to have slower speed. In containers we have zypper for now, but IMO we should not have any zypper in the end as we do not want containers to be updated in the image itself (unless we want customers to build their own proxy containers with some extensions).
On the proxy container host podman variant there is still zypper, but they can manage it the same way they do for regular proxy. We do not provide any hints.

My 2c only, I don't mean that this is wrong :)

**Comment by cbosdo** [comment]
> Hey @aaannz @cbosdo, there's my first take on how to allow proxy squid configuration.
>     * There's the field called `bandwidth_limit_kbps` in the `config.yaml`. It's added when the config is generated through the UI (see the screenshots above).

I wouldn't add anything more in the UI or API. Additional tuning of squid is an advanced feature and we don't want people to play with it without knowing what they do.

>     * The field is a positive integer, so if it's greater then 0, then the limit is set via the [delay pool](https://wiki.squid-cache.org/Features/DelayPools).
> 
>     * The delay pool config is a separate file and the entire `conf.d` dir [is added to the config](https://github.com/uyuni-project/uyuni/pull/6955/files#diff-786076059f6e2314df4d85db59705a8557a3d6531d6c2e51177c4277649dee78).
> 
>     * Whenever the container starts, the `containers/proxy-squid-image/uyuni-configure.py` script gets executed. It's setting the `/etc/squid/conf.d/delay.conf` depending on the proxy configuration `config.yaml`. [See here](https://github.com/uyuni-project/uyuni/pull/6955/files#diff-cfdad0aebaf1915e193f498d8ad1da4e1c0ccc4186a1c161448d3046072d6f10). **So basically we need a clean restart, so the container gets newly recreated and the script sets the config properly.**
> 
>     * for the systemd case there's an extension of the `containers/proxy-systemd-services/uyuni-proxy.sh ` [which is first mentioned here](https://documentation.suse.com/suma/4.3/en/suse-manager/installation-and-upgrade/proxy-container-installation.html#_using_a_custom_container_image_for_a_service). [It can replace](https://github.com/uyuni-project/uyuni/pull/6955/files#diff-ac8ef6e888c2587bbf36be5bcd625bb5e2b44dcdf9dca14cf588eda906c87b33) the `bandwidth_limit_kbps` value in the `/etc/uyuni/proxy/config.yaml` and restart the service.

The `config.yaml` has to be placed by the user in `/etc/uyuni/proxy/` for the `podman` and `systemd` case... so no need to change the `uyuni-proxy.sh` script. Touching that YAML file manually needs the user to know what they do. This means we have to document the possible options there and in the tuning guide or the proxy setup doc.

As @aaannz  mentions this only bring one possible setup. On one hand this avoids users to directly tamper and possibly mess up with the squid config, but on the other hand it restricts the use cases... I fear mounting an arbitrary file in the squid `conf.d` would bring funny L3s. I have no clue what would be best here. Especially with the delay pools, it seems you can have many possible configurations.

> ### Now the questions
> 
> What do we want to do with the salt here? We can create an sls to call either the `uyuni-proxy.sh` or the `kubectl` based on the container runtime.

We don't deploy the proxy with Salt: this has to be automated by the user for now. So don't bother about it in this PR.

>     * Should we also measure the bandwidth ourselves?

This is user-specific code... not for us to do anything.

**Comment by avshiliaev** [comment]
@cbosdo @aaannz I've reverted the changes to the GUI and the uyuni-proxy CLI, I've also added some additional parameters related to the problem of slow network. Here's a piece of documentation I would add covering the main options to tweak and their effect. As usual, I ask for you feedback here.

- **`client_lifetime`**: Adjust the **`client_lifetime`** parameter to limit the duration of client connections. This helps in releasing resources quickly and can be beneficial for slow networks. Add the following line to your configuration:

```
client_lifetime 1 minute
```
[More](http://www.squid-cache.org/Doc/config/client_lifetime/)

- **`request_timeout`**: Set an appropriate timeout value for client requests. This value determines how long Squid waits for a response from the upstream server. If the network is slow, you may want to increase this timeout. Add the following line to your configuration:

```
request_timeout 1 minute
```
[More](http://www.squid-cache.org/Doc/config/request_timeout/)

- **`range_offset_limit`**: Adjust the **`range_offset_limit`** parameter to optimize the transfer of large files. This value specifies the maximum number of bytes transferred at once when handling range requests. Increasing this value may improve performance for slow networks. Add the following line to your configuration:

```
range_offset_limit 256 KB
```
[More](http://www.squid-cache.org/Doc/config/range_offset_limit/)

- **`read_ahead_gap`**: Modify the **`read_ahead_gap`** parameter to control how much data Squid buffers before delivering it to clients. Increasing this value may reduce the number of network reads, improving performance on slow networks. Add the following line to your configuration:

```
read_ahead_gap 32 KB
```
[More](http://www.squid-cache.org/Doc/config/read_ahead_gap/)

- `bandwidth_limit_kbps` Delay pools feature in Squid can help in managing and controlling network bandwidth for clients, which can be beneficial for slow networks. The **`delay_pool`** feature allows you to implement traffic shaping and bandwidth control based on various factors such as client IP address, network segment, or other criteria. Currently we support one delay pool and you can control it by changing the `bandwidth_limit_kbps` option.

```
bandwidth_limit_kbps 256
```
[More](http://www.squid-cache.org/Doc/config/delay_pools/)

- **`client_req_buffer`** Squid buffers client requests by default, allowing it to process them efficiently. However, for slow networks, you can adjust the buffer size to optimize performance. In the Squid configuration file, you can modify the **`client_req_buffer`** directive. For example:

```
client_request_buffer_max_size 8 KB
```
[More](http://www.squid-cache.org/Doc/config/client_request_buffer_max_size/)


Remember to restart Squid after making any changes to the configuration. Monitor the performance after each adjustment to determine the optimal values for your specific slow network environment.

File: containers/proxy-squid-image/uyuni-configure.py [hunk 1]
```diff
@@ -2,17 +2,73 @@
...
 import re
+import shutil
+
 import yaml
 
+
+def set_cache_parameters(content, squid_config):
+    # max_cache_size_mb
+    content = re.sub(r"cache_dir aufs .*",
+                     f"cache_dir aufs /var/cache/squid {str(squid_config['max_cache_size_mb'])} 16 256",
```
Comment (cbosdo) [inline containers/proxy-squid-image/uyuni-configure.py:13]: We need to be able to read the old way of setting that value in the config file for compatibility reason.

File: containers/proxy-helm/templates/config.yaml [hunk 2]
```diff
@@ -9,8 +9,15 @@ data:
     server_version: "{{ .Values.server_version }}"
     proxy_fqdn: {{ .Values.proxy_fqdn }}
     email: {{ .Values.email }}
-    max_cache_size_mb: {{ .Values.max_cache_size_mb }}
```
Comment (cbosdo) [inline containers/proxy-helm/templates/config.yaml:12]: This still needs to be understood by the container config as there could be containers with this config in the wild.
The other solution is to not group the squid options under `squid_config`. Not sure which is best.

File: containers/proxy-squid-image/Dockerfile [hunk 3]
```diff
@@ -65,4 +65,4 @@ EXPOSE 8080/tcp
 
 VOLUME [ "/etc/uyuni", "/var/cache/squid" ]
 
-CMD uyuni-configure.py && squid -z --foreground && squid -FC -d 1 --foreground
+CMD uyuni-configure.py && squid -z --foreground && squid -FC -d 1 --foreground --enable-delay-pools
```
Comment (cbosdo) [inline containers/proxy-squid-image/Dockerfile:68]: Do we want to enable this option for everyone?

File: containers/proxy-squid-image/squid.conf [hunk 4]
```diff
@@ -2,6 +2,8 @@
 # To be used for Spacewalk Proxy servers.
 #
 
+/etc/squid/conf.d/*
```
Comment (cbosdo) [inline containers/proxy-squid-image/squid.conf:5]: Nice! the good thing is that the user could mount their own config volume there. This is not made easy with the systemd services but it's possible.

File: containers/proxy-squid-image/squid.conf [hunk 5]
```diff
@@ -73,3 +75,31 @@ range_offset_limit none
...
+
+# Delay pool 2
+# Do not limit bandwidth on local network(s) for any contents that have already been cached onto the proxy
+acl myendpoints1 src 192.168.1.0/24
+acl myendpoints2 src 192.168.122.0/24
+delay_class 2 2
+delay_access 2 allow myendpoints1 myendpoints2
+delay_access 2 deny all
+# -1/-1 means that there are no limits
+delay_parameters 2 -1/-1 -1/-1
```
Comment (cbosdo) [inline containers/proxy-squid-image/squid.conf:105]: Do we want to hard code those delay pools for everyone?

File: containers/proxy-squid-image/Dockerfile (hunk 3 above)
Comment (avshiliaev) [reply containers/proxy-squid-image/Dockerfile:68]: We could enable it conditionally I think, if the bandwidth limit is set. Let's see

File: containers/proxy-helm/templates/config.yaml (hunk 2 above)
Comment (avshiliaev) [reply containers/proxy-helm/templates/config.yaml:12]: Makes sense, I've removed the nesting of values

File: containers/proxy-squid-image/Dockerfile [hunk 6]
```diff
@@ -65,4 +65,4 @@ EXPOSE 8080/tcp
 
 VOLUME [ "/etc/uyuni", "/var/cache/squid" ]
 
-CMD uyuni-configure.py && squid -z --foreground && squid -FC -d 1 --foreground
+CMD configure_and_run.py
```
Comment (avshiliaev) [inline containers/proxy-squid-image/Dockerfile:68]: I've decided to wrap the call to squid into the py script. It's the easiest way to set the `--enable-delay-pools` flag conditionally without messing with env variables etc.

File: containers/proxy-squid-image/configure_and_run.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+            return f.read()
+
+    def write_config(self, squid_config):
+        with open(self.squid_conf_path, "w") as f:
+            f.seek(0, 0)
+            f.write(squid_config)
+            f.truncate()
+
+
+class SquidConfigurator:
```
Comment (avshiliaev) [inline containers/proxy-squid-image/configure_and_run.py:31]: The config script is getting more and more complex to I've decided to rewrite it in a slightly more OOP way and added some simple unittests to make it a bit more transparent

**Comment by cbosdo** [comment]
Still needs discussion

**Pull request closed without merging**